oap-server/analyzer/event-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/event/listener/EventRecordAnalyzerListener.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.analyzer.event.listener;
+
+import com.google.gson.Gson;
+import lombok.RequiredArgsConstructor;
+import org.apache.skywalking.apm.network.event.v3.Source;
+import org.apache.skywalking.oap.server.core.CoreModule;
+import org.apache.skywalking.oap.server.core.analysis.TimeBucket;
+import org.apache.skywalking.oap.server.core.analysis.worker.MetricsStreamProcessor;
+import org.apache.skywalking.oap.server.core.config.NamingControl;
+import org.apache.skywalking.oap.server.core.event.Event;
+import org.apache.skywalking.oap.server.library.module.ModuleManager;
+
+/**
+ * EventRecordAnalyzerListener forwards the event data to the persistence layer with the query required conditions.
+ */
+@RequiredArgsConstructor
+public class EventRecordAnalyzerListener implements EventAnalyzerListener {
+    private static final Gson GSON = new Gson();
+
+    private final NamingControl namingControl;
+
+    private final Event event = new Event();
+
+    @Override
+    public void build() {
+        MetricsStreamProcessor.getInstance().in(event);
+    }
+
+    @Override
+    public void parse(final org.apache.skywalking.apm.network.event.v3.Event e) {
+        event.setUuid(e.getUuid());
+
+        if (e.hasSource()) {
+            final Source source = e.getSource();
+            event.setService(namingControl.formatServiceName(source.getService()));
+            event.setServiceInstance(namingControl.formatInstanceName(source.getServiceInstance()));
+            event.setEndpoint(namingControl.formatEndpointName(source.getService(), source.getEndpoint()));
+        }
+
+        event.setName(e.getName());
+        event.setType(e.getType().name());
+        event.setMessage(e.getMessage());
+        event.setParameters(GSON.toJson(e.getParametersMap()));
+        event.setStartTime(e.getStartTime());
+        event.setEndTime(e.getEndTime());
+        event.setTimeBucket(TimeBucket.getMinuteTimeBucket(e.getStartTime()));
```

Comment:
Do you have start time when reporting end time? I remember you and @BFergerson discussed this, this should be only optional. And from my understanding, it isn't just optional, it is immutable.

Reply:
> Do you have start time when reporting end time? I remember you and @BFergerson discussed this, this should be only optional.

No, also that's not mandatory. It's optional when reporting an `endTime`.

> And from my understanding, it isn't just optional, it is immutable.

It's indeed immutable, we won't touch it when merging two records, only `endTime` and `message` (`parameters`) will be merged

Reply:
But what happens if someone reports an `end` event without `start` event`? So someone, the `start` event lost or blocked in some channel?

Reply:
> But what happens if someone reports an end event without start event? Such as, the start event lost or blocked in some channel?

Do you think we should tackle the case where the `end` event comes before the `start` event, (i.e. the `start` and `end` are unordered), or simply abandon the `end` event if there is no `start` event already?

Reply:
Usually `end` should be later at least several seconds before `start`, but it may be very close. Such as rebooting could be quickly enough. 
The abandon mechanism is the thing we should discuss, where to do. In my mind, there are 2 places
1. Query before sending the metrics into the stream.
2. Adjust merge return value (`void`->`boolean`) to support `abandon` at the last minute. But this would cause wider changes

Reply:
Method 2 seems to be more elegant. What about adjusting the signature of `combine` like this?

```java
    /**
     * Merge the given metrics instance, these two must be the same metrics type.
     *
     * @param metrics to be merged
     * @return {@code true} if the combined metrics should be continuously processed, {@code false} it should be abandoned.
     */
    public abstract boolean combine(Metrics metrics);
```

Reply:
Could you submit another PR to change that first? This change makes sense to me, but will affect a lot of codes, I hope the commit log is clear to everyone.

Reply:
> Could you submit another PR to change that first? This change makes sense to me, but will affect a lot of codes, I hope the commit log is clear to everyone.

https://github.com/apache/skywalking/pull/6318